File: PRACTICA#3/GENERACIDAD/1/JAVA/Caja.java
import java.util.Optional;
//a)
public class Caja<T> {
    private Optional<T> contenido;

    public Caja() {
        contenido = Optional.empty();
    }
//a)
    public void guardar(T objeto) {
        contenido = Optional.of(objeto);
        System.out.println("Objeto guardado en la caja: " + objeto);
    }

    public Optional<T> obtener() {
        if (contenido.isPresent()) {
            System.out.println("Obteniendo objeto de la caja: " + contenido.get());
            return contenido;
        } else {
            System.out.println("La caja está vacía");
            return Optional.empty();
        }
    }

    public boolean estaVacia() {
        return !contenido.isPresent();
    }

    @Override
    public String toString() {
        return contenido.isPresent() 
            ? "Caja[" + contenido.get().getClass().getSimpleName() + "]: " + contenido.get()
            : "Caja vacía";
    }
}

File: PRACTICA#3/GENERACIDAD/1/PYTHON/Caja.py
from typing import TypeVar, Generic, Optional

T = TypeVar('T')

class Caja(Generic[T]):
    
    def __init__(self):
        self._contenido: Optional[T] = None
    
    #a)
    def guardar(self, objeto: T) -> None:
        self._contenido = objeto
        print(f"Objeto guardado en la caja: {objeto}")
    
    def obtener(self) -> Optional[T]:
        if self._contenido is not None:
            print(f"Obteniendo objeto de la caja: {self._contenido}")
            return self._contenido
        else:
            print("La caja está vacía")
            return None
    
    def esta_vacia(self) -> bool:
        return self._contenido is None
    
    def __str__(self) -> str:
        if self._contenido is not None:
            return f"Caja[{type(self._contenido).__name__}]: {self._contenido}"
        else:
            return "Caja vacía"